### PR TITLE
Add support for RwLock

### LINKED ISSUE #21: Add support for RwLock
I notice that `loom::sync` has a [mock implementation of `std::sync::Mutex`](https://github.com/carllerche/loom/blob/master/src/sync/mutex.rs) but lacks a mock `std::sync::RwLock`. It would be nice to be able to test code using read-write locks with `loom`.

### PR DESCRIPTION
**Pull request by pop**
Fixes #21 

Oh snap when did this move to the Tokio org? Well, all the same: I added support for RwLock!

I don't know if this is done, but the tests (which I got from the RwLock docs) pass so I'm happy.

I would love some feedback. The `is_read_locked()` and `is_write_locked()` functions cause "Already Locked" panics so I am not using them at the moment. Let me know if you want more info on that.

cc: @pickfire ?

### CODE REVIEW

File: src/rt/rwlock.rs [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+            RwLock { obj }
+        })
+    }
+
+    fn get_state<'a>(&self, objects: &'a mut object::Store) -> &'a mut State {
+        self.obj.rwlock_mut(objects).unwrap()
+    }
+}
+
+impl RwLock {
```
Comment (pickfire) [inline src/rt/rwlock.rs]: I wonder why is this separated from the `RwLock` impl above?

**Comment by pickfire** [comment]
@pop Nice, I believe I may be lack of experience to review this but probably @hawkw could help.

File: .gitignore [hunk 2]
```diff
@@ -1,2 +1,6 @@
 target
 Cargo.lock
+
+# Vim
+.*.swp
+.*.swo
```
Comment (hawkw) [inline .gitignore]: Nit: ideally, we'd add this in a separate PR so it's not part of the diff for adding the mock `RwLock`.

File: src/rt/rwlock.rs [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+use std::collections::HashSet;
+use std::sync::atomic::Ordering::{Acquire, Release};
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub(crate) struct RwLock {
+    obj: Object,
+}
+
+#[derive(Debug, PartialEq)]
+enum RwLockType {
```
Comment (hawkw) [inline src/rt/rwlock.rs]: Nit: i'm not a huge fan of this naming...but I'm not really sure what would be better off the top of my head. Maybe `Locked`?

File: src/rt/rwlock.rs [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+    Write(thread::Id),
+}
+
+#[derive(Debug)]
+pub(super) struct State {
+    /// If the rwlock should establish sequential consistency.
+    seq_cst: bool,
+
+    /// `Some` when the rwlock is in the locked state.
+    /// The `thread::Id` references the thread that currently holds the rwlock.
```
Comment (hawkw) [inline src/rt/rwlock.rs]: I think this comment could be made clearer, since the `lock` value could either be a single thread ID holding the write lock, or multiple thread IDs holding read locks.

File: src/rt/rwlock.rs [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+                lock: None,
+                last_access: None,
+                synchronize: Synchronize::new(execution.max_threads),
+            });
+
+            RwLock { obj }
+        })
+    }
+
+    fn get_state<'a>(&self, objects: &'a mut object::Store) -> &'a mut State {
```
Comment (hawkw) [inline src/rt/rwlock.rs]: Nit/style: I'd prefer it if we organized this impl block with all the `pub(crate)` methods first, followed by the private methods...not going to block merging the PR on that, though.

File: src/rt/rwlock.rs [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+        super::execution(
+            |execution| match self.get_state(&mut execution.objects).lock {
+                Some(RwLockType::Read(_)) => true,
+                _ => false,
+            },
+        )
+    }
+}
+
+impl RwLock {
```
Comment (hawkw) [inline src/rt/rwlock.rs]: Why are there multiple `impl` blocks here? Can these be combined into one impl block?

File: src/rt/rwlock.rs [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+                None => {
+                    let mut threads: HashSet<thread::Id> = HashSet::new();
+                    threads.insert(thread_id);
+                    threads
+                }
+                Some(RwLockType::Read(current)) => {
+                    let mut threads: HashSet<thread::Id> = HashSet::new();
+                    threads.extend(current);
+                    threads.insert(thread_id);
+                    threads
```
Comment (hawkw) [inline src/rt/rwlock.rs]: Is there a reason we create a new `HashSet` here, rather than something like this:
```rust
let thread_set = match state.lock {
    // ...
    Some(RwLockType::Read(ref mut current)) => {
        current.insert(thread_id);
        current
    },
    // ...
}

File: src/rt/rwlock.rs [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+            };
+
+            dbg!(state.synchronize.sync_load(&mut execution.threads, Acquire));
+
+            if state.seq_cst {
+                // Establish sequential consistency between locks
+                execution.threads.seq_cst();
+            }
+
+            // Block all other threads attempting to acquire rwlock
```
Comment (hawkw) [inline src/rt/rwlock.rs]: Nit/TIOLI: seems like this code is repeated enough that it could maybe be factored out into its own function...not a blocker.

File: src/rt/rwlock.rs [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+
+            true
+        })
+    }
+
+    fn is_write_locked(&self) -> bool {
+        super::execution(
+            |execution| match self.get_state(&mut execution.objects).lock {
+                Some(RwLockType::Write(_)) => true,
+                Some(RwLockType::Read(_)) | None => false,
```
Comment (hawkw) [inline src/rt/rwlock.rs]: TIOLI:
```suggestion
                _ => false,
```

File: src/rt/rwlock.rs [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+    /// Tracks write access to the rwlock.
+    last_access: Option<Access>,
+
+    /// Causality transfers between threads
+    synchronize: Synchronize,
+}
+
+impl RwLock {
+    /// Common RwLock function
+    pub(crate) fn new(seq_cst: bool) -> RwLock {
```
Comment (hawkw) [inline src/rt/rwlock.rs]: As far as I can tell, the `rt::RwLock` type is only ever constructed in `sync::RwLock::new`, and it always passes `true` here. Will we ever construct a `rt::RwLock` with `seq_cst=false`? If not, should we just remove the bool and always establish sequential consistency?

File: src/sync/rwlock.rs [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+pub struct RwLock<T> {
+    object: rt::RwLock,
+    data: std::sync::RwLock<T>,
+}
+
+/// Mock implementation of `std::sync::RwLockReadGuard`
+#[derive(Debug)]
+pub struct RwLockReadGuard<'a, T> {
+    lock: &'a RwLock<T>,
+    data: Option<std::sync::RwLockReadGuard<'a, T>>,
```
Comment (hawkw) [inline src/sync/rwlock.rs]: It's not immediately obvious why this is an `Option`. I think it's it so that the `Drop` impl can drop the `std` guard and release the `std` lock _before_ releasing the `loom` mock lock, because that might cause another thread to acquire the lock? 

I think a comment explaining this (either here on the field or in the `Drop` impl) would be very helpful.

File: src/rt/rwlock.rs (hunk 6 above)
Comment (pop) [reply src/rt/rwlock.rs]: It was just for readability. One block is the reading methods, one block is the writing methods, third for the rest. I can combine them 😄

File: src/rt/rwlock.rs (hunk 7 above)
Comment (pop) [reply src/rt/rwlock.rs]: I didn't go down that route because it quickly unraveled into needing to specify lifetime parameters everywhere and I am not adept enough at Rust to do that... with grace. I'll dig into it though and see if I can make it work.

File: src/rt/rwlock.rs (hunk 5 above)
Comment (pickfire) [reply src/rt/rwlock.rs]: Shouldn't it be called `state` according to API guidelines?

File: src/rt/rwlock.rs (hunk 6 above)
Comment (pickfire) [reply src/rt/rwlock.rs]: Oh, I was also asking about that in another commit.

File: src/rt/rwlock.rs (hunk 5 above)
Comment (pop) [reply src/rt/rwlock.rs]: An identical method is called `get_state` in Mutex FWIW.

**Comment by pop** [comment]
I'm taking a second look at some of what I implemented. In trying to factor out some code I think I found a bug in my implementation. I'll remove the WIP title when this is ready for review again.

File: src/rt/rwlock.rs (hunk 7 above)
Comment (hawkw) [reply src/rt/rwlock.rs]: Hmm, if it isn't easily changed, I don't think this is a blocker, it just seems kind of inefficient...

**Comment by pop** [comment]
This is ready for a final review... or at least I think it is right now... so throw your feedback at me (again)!

Some feedback I haven't taken:
* A part of the code looks like I should be able to edit a HashMap but instead I create a new one. I need a bit more hand-holding to get that refactor done. I don't know how to make rustc happy. When I end up putting lifetime specifiers in stuff that is three steps removed from the change I'm making I start to think "this is not a good refactor".
* There's some repeated code. I refactored out that loop and a similar loop for symmetry. I'm not super sure about that choice, but the loops had a bug so I figured putting them close to eachother in the code might make it easier to contrast the two.

**Comment by hawkw** [comment]
@pop re:
> A part of the code looks like I should be able to edit a HashMap but instead I create a new one. I need a bit more hand-holding to get that refactor done. I don't know how to make rustc happy. 

what do you think about something like this:
```rust
            // Set the lock to the current thread
            let mut already_locked = false;
            state.lock = match state.lock.take() {
                None => {
                    let mut threads: HashSet<thread::Id> = HashSet::new();
                    threads.insert(thread_id);
                    Some(Locked::Read(threads))
                }
                Some(Locked::Read(threads)) => {

                    threads.insert(thread_id);
                    Some(Locked::Read(threads))
                }
                Some(Locked::Write(writer)) => {
                    already_locked = true;
                    Some(Locked::Write(writer))
                }
            };

            if already_locked {
                return false;
            }
```

File: src/rt/rwlock.rs [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+enum Locked {
+    Read(HashSet<thread::Id>),
+    Write(thread::Id),
+}
+
+#[derive(Debug)]
+pub(super) struct State {
+    /// A single `thread::Id` when Write locked.
+    /// A set of `thread::Id` when Read locked.
+    lock: Option<Locked>,
```
Comment (pickfire) [inline src/rt/rwlock.rs]: Why does this have to be `Option<Locked>`? Could it be `Locked` and initialized at first use?

Comment (pop) [reply src/rt/rwlock.rs]: If no threads are reading or writing the RwLock then it doesn't make sense for it to be in either `Locked::Read` or `Locked::Write`, so I think `None` makes the most sense.

**Comment by pop** [comment]
Thanks @hawkw and @pickfire for the review(s)! I'm learning a lot :smile:

File: src/rt/rwlock.rs (hunk 12 above)
Comment (pickfire) [reply src/rt/rwlock.rs]: Ah, so similar to the following
```rust
#[derive(Debug, PartialEq)]
enum Locked {
    Read(HashSet<thread::Id>),
    Write(thread::Id),
    None,
}
```
I guess `Option<Locked>` makes more sense now but I wonder if the size is any different from `Option<Locked>` compared to `Locked`.

File: src/rt/object.rs [hunk 13]
```diff
@@ -106,6 +107,15 @@ impl Object {
             _ => None,
         }
     }
+
+    pub(super) fn rwlock_mut(self, store: &mut Store) -> Option<&mut rwlock::State> {
+        assert_eq!(self.execution_id, store.execution_id);
+
+        match &mut store.entries[self.index] {
```
Comment (pickfire) [inline src/rt/object.rs]: Why is the `&mut` required here? I rarely sees a `&mut` for match statement, I wonder what it does.

File: tests/mutex.rs [hunk 14]
```diff
@@ -1,4 +1,4 @@
-#![deny(warnings, rust_2018_idioms)]
+// #![deny(warnings, rust_2018_idioms)]
```
Comment (pickfire) [inline tests/mutex.rs]: Why is this change necessary?

File: src/rt/rwlock.rs (hunk 12 above)
Comment (pop) [reply src/rt/rwlock.rs]: Hmm. I don't know about that. I was more focused on the expressiveness of the implementation and not the size impact. I can change it to your suggestion if we're concerned about it. Your suggestion seems pretty expressive too -- "the RwLock can be in one of three states..." makes sense too.

File: src/rt/object.rs (hunk 13 above)
Comment (pop) [reply src/rt/object.rs]: In this case I was following the other `*_mut` methods. `mutext` does the same thing so I copy-pasted it. I can look into it, but I think consistency (all of the `*_mut` methods look _exactly_ the same) is important here so we should update all methods -- or I'll need to figure out why `rwlock` is different from `mut`.

Comment (pop) [reply src/rt/object.rs]: * To answer your question I think we `&mut` here because we are returning an `Option<&mut rwlock::State>` in the `match`. If I try to get rid of the `&mut` here I get messages like this:
```rs
error[E0308]: mismatched types
   --> src/rt/object.rs:115:38
    |
115 |             Entry::RwLock(v) => Some(v),
    |                                      ^
    |                                      |
    |                                      expected mutable reference, found struct `rt::rwlock::State`
    |                                      help: consider mutably borrowing here: `&mut v`
    |
    = note: expected type `&mut rt::rwlock::State`
               found type `rt::rwlock::State`
```

but I can't return `Some(&mut v)` because we're not borrowing as mutable.

File: tests/mutex.rs (hunk 14 above)
Comment (pop) [reply tests/mutex.rs]: It wasn't 😅 -- just a typo.

File: src/rt/rwlock.rs (hunk 5 above)
Comment (hawkw) [reply src/rt/rwlock.rs]: @pickfire I also thought about mentioning that, but I saw that `Mutex` had a function called `get_state`, and I thought consistency was good... 🤷‍♀

File: src/rt/rwlock.rs (hunk 12 above)
Comment (pickfire) [reply src/rt/rwlock.rs]: No, I am not sure whether it affects the size, just thought about it, we need to verify it with probably something like `std::mem::size_of<Type>()`.

File: src/rt/rwlock.rs [hunk 15]
```diff
@@ -182,22 +182,28 @@ impl RwLock {
...
-                    Some(Locked::Read(new))
+                Some(Locked::Read(mut threads)) => {
+                    threads.insert(thread_id);
+                    Some(Locked::Read(threads))
+                }
+                Some(Locked::Write(writer)) => {
+                    already_locked = true;
+                    Some(Locked::Write(writer))
                 }
-                Some(Locked::Write(_)) => return false,
```
Comment (pickfire) [inline src/rt/rwlock.rs]: Why not just keep
```rust
                 Some(Locked::Write(_)) => return false,
```
and ignore
```rust

                 Some(Locked::Write(writer)) => {
                    already_locked = true;
```

Comment (hawkw) [reply src/rt/rwlock.rs]: @pickfire because of the [call to `Option::take` on line 186](https://github.com/tokio-rs/loom/pull/88/commits/cd84f9dd31438c588f53ef507e50ac3b1f952efc#diff-e9106da480ca203b688aa13f79b66051R186) that would be incorrect. Since we've `take`n the previous value, `state.lock` is currently `None`. If we return false here, we leave the state as `None`, incorrectly releasing the lock.

**Comment by pop** [comment]
I think we're just waiting on a review from @carllerche, is that right @hawkw @pickfire?

**Comment by hawkw** [comment]
> I think we're just waiting on a review from @carllerche, is that right @hawkw @pickfire?
@pop as far as I know, that's right! This looks good to me!

**Comment by pickfire** [comment]
Yup, looks good to me too.

**Comment by pop** [comment]
Rebased against master so there shouldn't be any conflicts. Thanks @hawkw (via #105)!

**Comment by carllerche** [comment]
I merged master and fixed the conflicts :+1:

**Comment by carllerche** [comment]
Sorry for the delay, thanks for doing the work!

**Comment by carllerche** [comment]
Btw, there is a loom channel on the Tokio discord (https://discord.gg/tokio). If something is blocked on me, feel free to ping me there.
